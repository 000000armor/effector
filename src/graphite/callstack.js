//@flow

const printFull = (full, tag = '') => {
  const items = full.map(e => `${e.type} ${e.id}`)
  const itemsText = items.join(', ')
  console.log(`${tag} [${itemsText}]`)
}

export default {
  box: [],
  item: [],
  full: [],
  pushItem(item: *) {
    this.item.push(item)
    this.full.push(item)
    dev: {
      // printFull(this.full)
      // printFull(this.item, '(cmd)')
    }
  },
  popItem() {
    this.item.pop()
    this.full.pop()
    dev: {
      // printFull(this.full)
      // printFull(this.item, '(cmd)')
    }
  },
  pushBox(box: *) {
    this.box.push(box)
    this.full.push(box)
    dev: {
      // printFull(this.full)
      // printFull(this.box, '(box)')
    }
  },
  popBox() {
    this.box.pop()
    this.full.pop()
    dev: {
      // printFull(this.full)
      // printFull(this.box, '(box)')
    }
  },
}
