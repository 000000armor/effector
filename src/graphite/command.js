//@flow

import type {TypeDef} from 'effector/stdlib'
// import {__DEV__, __DEBUG__} from 'effector/flags'
import type {CommandList} from './index.h'
declare var __step: TypeDef<*, 'step'>
declare var __single: any
declare function __val(key: string, value: any): any

const cmd = {
  emit(meta, local) {},
  filter(meta, local) {
    const runCtx = tryRun({
      err: false,
      result: (null: any),
      arg: local.__stepArg,
      val: meta.val,
      fn: __single.filter,
    })
    local.isChanged = Boolean(runCtx.result)
  },
  run(meta, local) {
    if ('pushUpdate' in __single) {
      __single.pushUpdate = data => meta.pendingEvents.push(data)
    }
    const runCtx = tryRun({
      err: false,
      result: (null: any),
      arg: local.__stepArg,
      val: meta.val,
      fn: __single.runner,
    })
    local.isFailed = runCtx.err
  },
  update(meta, local) {
    local.store.current = local.__stepArg
  },
  compute(meta, local) {
    const runCtx = tryRun({
      err: false,
      result: (null: any),
      arg: local.__stepArg,
      val: meta.val,
      fn: __single.fn,
    })
    local.isChanged = !runCtx.err
    ///TODO WARNING!! DIRTY HACK REMOVE ASAP
    ///need to separate pre and post local variables
    local.__stepArg = runCtx.err ? null : runCtx.result
  },
}

const transition = {
  emit: () => true,
  filter: (meta, local) => local.isChanged,
  run: (meta, local) => local.isFailed,
  update: () => true,
  compute: (meta, local) => local.isChanged,
}

const local = {
  emit: meta => ({__stepArg: __val('scope').top.__stepArg}),
  filter: meta => ({__stepArg: __val('scope').top.__stepArg, isChanged: false}),
  run: meta => ({__stepArg: __val('scope').top.__stepArg, isFailed: true}),
  update: meta => ({
    __stepArg: __val('scope').top.__stepArg,
    store: __single.store,
  }),
  compute: meta => ({
    __stepArg: __val('scope').top.__stepArg,
    isChanged: false,
  }),
}

export default ({
  emit: {
    cmd: cmd.emit,
    transition: transition.emit,
    local: local.emit,
  },
  filter: {
    cmd: cmd.filter,
    transition: transition.filter,
    local: local.filter,
  },
  run: {
    cmd: cmd.run,
    transition: transition.run,
    local: local.run,
  },
  update: {
    cmd: cmd.update,
    transition: transition.update,
    local: local.update,
  },
  compute: {
    cmd: cmd.compute,
    transition: transition.compute,
    local: local.compute,
  },
}: CommandList)

const tryRun = ctx => {
  try {
    ctx.result = ctx.fn.call(null, ctx.arg, ctx.val)
  } catch (err) {
    console.error(err)
    ctx.err = true
  }
  return ctx
}
