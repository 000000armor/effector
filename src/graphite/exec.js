//@flow

import type {TypeDef, Graph} from 'effector/stdlib'
import type {Meta, CommandList} from './index.h'

type Line = {|
  +step: TypeDef<*, 'step'>,
  +pre: Array<PostAction>,
  +post: Array<PostAction>,
|}

type Area = {|
  +list: $ReadOnlyArray<Line>,
  +pre: Array<PostAction>,
  +post: Array<PostAction>,
  index: number,
|}

type PostAction =
  | {|+type: 'meta/?stop'|}
  | {|+type: 'meta/!stop'|}
  | {|+type: 'head/--'|}
  | {|+type: 'callstack/pop'|}
  | {|+type: 'scope/size', +size: number|}
  | {|+type: 'scope/push', +scope: {+arg: any}|}

const step = (meta: Meta): TypeDef<*, 'step'> =>
  meta.callstack[meta.callstack.length - 1]
const single = (meta: Meta) => step(meta).data.data

export function exec(unit: {+graphite: Graph<any>, ...}, payload: any) {
  runtime(unit.graphite, payload)
}
export function runtime(graph: Graph<any>, payload: any) {
  const meta: Meta = {
    callstack: [],
    pendingEvents: [],
    stop: false,
    scope: [
      {
        arg: payload,
      },
    ],
    val: {},
  }
  runStep(graph.seq, meta)
  runPendings(meta.pendingEvents)
}
const runPendings = pendings => {
  if (pendings.length === 0) return
  const ordered = []
  const orderedIDs = []
  for (let i = pendings.length - 1; i >= 0; i--) {
    const item = pendings[i]
    if (orderedIDs.indexOf(item.event.id) !== -1) continue
    orderedIDs.push(item.event.id)
    ordered.push(item)
  }
  if (ordered.length === 0) return
  for (let i = ordered.length - 1; i >= 0; i--) {
    const item = ordered[i]
    item.event(item.data)
  }
}
const runStep = (step: TypeDef<*, 'step'>, meta: Meta) => {
  const scope = meta.scope
  //prettier-ignore
  const pos: {|
    head: number,
    +stack: Array<Area>,
  |} = {
    head: 0,
    stack: [{
      list: [{
        step,
        pre: [],
        post: [
          {type: 'callstack/pop'},
        ],
      }],
      pre: [],
      post: [],
      index: 0,
    }],
  }
  headLoop: while (pos.head >= 0) {
    const area = pos.stack[pos.head]
    runActions(area.pre, area)
    while (area.index < area.list.length) {
      const line = area.list[area.index]
      const step = line.step
      meta.stop = false
      meta.callstack.push(step)
      runActions(line.pre, area)
      switch (step.type) {
        case 'single': {
          const type = step.data.type
          const local = command[type].local(meta, scope[scope.length - 1].arg)
          scope.push(local)
          command[type].cmd(meta, local)
          meta.stop = !command[type].transition(meta, local)
          break
        }
        case 'multi': {
          line.post.push({type: 'meta/!stop'})
          const items = step.data
          const list: Array<Line> = Array(items.length)
          for (let i = 0; i < items.length; i++) {
            list[i] = {
              step: items[i],
              pre: [],
              post: [
                {type: 'callstack/pop'},
                {type: 'scope/size', size: scope.length},
              ],
            }
          }
          area.index += 1
          pos.head += 1
          pos.stack[pos.head] = {
            list,
            pre: [],
            post: line.post,
            index: 0,
          }
          continue headLoop
        }
        case 'seq': {
          line.post.push({type: 'scope/size', size: scope.length})
          const items: Array<TypeDef<*, 'step'>> = step.data
          const list: Array<Line> = Array(items.length)
          for (let i = 0; i < items.length; i++) {
            //prettier-ignore
            list[i] = {
              step: items[i],
              pre: [],
              post: [
                {type: 'callstack/pop'},
                {type: 'meta/?stop'},
              ],
            }
          }
          area.index += 1
          pos.head += 1
          pos.stack[pos.head] = {
            list,
            pre: [],
            post: line.post,
            index: 0,
          }
          continue headLoop
        }
      }
      runActions(line.post, area)
      area.index += 1
    }
    runActions(area.post, area)
    pos.head -= 1
  }

  function runActions(post: Array<PostAction>, area: Area) {
    for (let i = post.length - 1; i >= 0; i--) {
      const action = post[i]
      switch (action.type) {
        case 'meta/?stop':
          if (meta.stop) area.index = Infinity
          break
        case 'head/--':
          pos.head -= 1
          break
        case 'callstack/pop':
          meta.callstack.pop()
          break
        case 'meta/!stop':
          meta.stop = false
          break
        case 'scope/size':
          scope.length = action.size
          break
        case 'scope/push':
          scope.push(action.scope)
          break
        /*::
        default:
          (action.type: empty)
        */
      }
    }
  }
}

const cmd = {
  emit(meta, local) {},
  filter(meta, local) {
    const runCtx = tryRun({
      err: false,
      result: (null /*: any*/),
      arg: local.arg,
      val: meta.val,
      fn: single(meta).fn,
    })
    local.isChanged = Boolean(runCtx.result)
  },
  run(meta, local) {
    const data = single(meta)
    if ('pushUpdate' in data) {
      data.pushUpdate = data => meta.pendingEvents.push(data)
    }
    const runCtx = tryRun({
      err: false,
      result: (null /*: any*/),
      arg: local.arg,
      val: meta.val,
      fn: data.fn,
    })
    local.isFailed = runCtx.err
  },
  update(meta, local) {
    local.store.current = local.arg
  },
  compute(meta, local) {
    const runCtx = tryRun({
      err: false,
      result: (null /*: any*/),
      arg: local.arg,
      val: meta.val,
      fn: single(meta).fn,
    })
    local.isChanged = !runCtx.err
    ///TODO WARNING!! DIRTY HACK REMOVE ASAP
    ///need to separate pre and post local variables
    local.arg = runCtx.err ? null : runCtx.result
  },
}

const command = ({
  emit: {
    cmd: cmd.emit,
    transition: () => true,
    local: (meta, arg) => ({arg}),
  },
  filter: {
    cmd: cmd.filter,
    transition: (meta, local) => local.isChanged,
    local: (meta, arg) => ({arg, isChanged: false}),
  },
  run: {
    cmd: cmd.run,
    transition: (meta, local) => local.isFailed,
    local: (meta, arg) => ({arg, isFailed: true}),
  },
  update: {
    cmd: cmd.update,
    transition: () => true,
    local: (meta, arg) => ({
      arg,
      store: single(meta).store,
    }),
  },
  compute: {
    cmd: cmd.compute,
    transition: (meta, local) => local.isChanged,
    local: (meta, arg) => ({
      arg,
      isChanged: false,
    }),
  },
}: CommandList)

const tryRun = ctx => {
  try {
    ctx.result = ctx.fn.call(null, ctx.arg, ctx.val)
  } catch (err) {
    console.error(err)
    ctx.err = true
  }
  return ctx
}
