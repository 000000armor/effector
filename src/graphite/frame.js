//@flow

import {atom, type Atom} from '../effector/atom'

import {step as Step} from 'effector/datatype/FullDatatype.bs'

export const seq: Atom<Step.Seq | null> = atom(null)
export const frame = (atom(null): Atom<Set<Step.Step> | null>)
