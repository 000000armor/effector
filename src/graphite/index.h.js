//@flow

import type {TypeDef, StateRef} from 'effector/stdlib'

export type CtxKind =
  | 'compute'
  | 'emit'
  | 'filter'
  | 'update'
  | 'run'
  | 'combine'
export type CommonCtx = {arg: any, ...}

export type Meta = {|
  callstack: Array<TypeDef<*, *>>,
  stop: boolean,
  scope: Array<any>,
  pendingEvents: Array<{
    event: (data: any) => any,
    data: any,
  }>,
  val: {[name: string]: StateRef},
|}

export type Command<Local> = {
  cmd(meta: Meta, local: Local): void,
  transition(meta: Meta, local: Local): boolean,
  local(meta: Meta): Local,
}

export type CommandList = $ReadOnly<{
  emit: Command<{arg: any}>,
  filter: Command<{arg: any, isChanged: boolean}>,
  run: Command<{arg: any, isFailed: boolean}>,
  update: Command<{arg: any, store: StateRef}>,
  compute: Command<{arg: any, isChanged: boolean}>,
}>

export type StepVisitor = (meta: Meta) => void

export type Line = {|
  +step: TypeDef<*, 'step'>,
  +pre: Array<PostAction>,
  +post: Array<PostAction>,
|}

export type Area = {|
  +list: $ReadOnlyArray<Line>,
  +pre: Array<PostAction>,
  +post: Array<PostAction>,
  index: number,
  active: boolean,
|}
export type Pos = {|
  head: number,
  +stack: Array<Area>,
|}
export type PostAction =
  | {|+type: 'meta/?stop'|}
  | {|+type: 'meta/!stop'|}
  | {|+type: 'head/--'|}
  | {|+type: 'callstack/pop'|}
  | {|+type: 'scope/size', +size: number|}
  | {|+type: 'scope/push', +scope: {+arg: any}|}
