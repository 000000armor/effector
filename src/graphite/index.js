//@flow

export {walkEvent} from './walk'
