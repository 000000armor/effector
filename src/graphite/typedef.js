//@flow
import {typeDef} from 'effector/stdlib/typedef'

const type = () => null

export const Step = typeDef(('step': 'step'), {
  single: type(),
  multi: type(),
  seq: type(),
})

export const Cmd = typeDef(('cmd': 'cmd'), {
  compute: type(),
  emit: type(),
  run: type(),
  filter: type(),
  update: type(),
})

export const Ctx = typeDef(('ctx': 'ctx'), {
  compute: (args, result, error, isError, isNone, isChanged) => ({
    args,
    result,
    error,
    isError,
    isNone,
    isChanged,
  }),
  emit: (eventName, payload) => ({eventName, payload}),
  run: (args, parentContext) => ({args, parentContext}),
  filter: (value, isChanged) => ({value, isChanged}),
  update: value => ({value}),
})
