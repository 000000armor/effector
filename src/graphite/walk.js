//@flow
import invariant from 'invariant'
import type {Event} from 'effector/event'
import type {StateRef} from 'effector/stdlib/stateref'
import {Ctx} from 'effector/graphite/typedef'
import type {TypeDef} from 'effector/stdlib/typedef'

import callstack from './callstack'

export function walkEvent<T>(payload: T, event: Event<T>) {
  const steps: TypeDef<'seq', 'step'> = event.graphite.seq
  const eventCtx = Ctx.emit({
    eventName: event.getType(),
    payload,
  })
  walkNode(steps, eventCtx)
}

type CommonCtx = TypeDef<'compute' | 'emit' | 'filter' | 'update', 'ctx'>
type Meta = {
  ctx: TypeDef<'compute' | 'emit' | 'filter' | 'update', 'ctx'>,
  stop: boolean,
  transactions: Array<() => void>,
  arg: any,
}
export function walkNode(seq: TypeDef<'seq', 'step'>, ctx: TypeDef<*, 'ctx'>) {
  const meta = {
    transactions: ([]: Array<() => void>),
    stop: false,
    ctx,
  }
  runStep(seq, ctx, meta)
  for (let i = 0; i < meta.transactions.length; i++) {
    meta.transactions[i]()
  }
  meta.transactions.length = 0
}
function runStep(step, ctx: *, meta) {
  invariant(step.type in stepVisitor, 'impossible case "%s"', step.type)
  callstack.pushBox(step)
  meta.stop = false
  stepVisitor[step.type](step, ctx, meta)
  callstack.popBox()
}
const stepVisitor = {
  choose(step: TypeDef<'choose', 'step'>, ctx: TypeDef<*, 'ctx'>, meta) {
    const ref: StateRef = step.data.ref
    const selector: TypeDef<*, 'step'> = step.data.selector
    const cases: {+[key: string]: TypeDef<*, 'step'>} = step.data.cases
    runStep(selector, ctx, meta)
    const caseName = ref.current
    //optional
    invariant(
      typeof caseName === 'string',
      'incorrect selector "%s" for id %s',
      caseName,
      ref.id,
    )
    let next
    if (caseName in cases) {
      next = cases[caseName]
    } else if ('__' in cases) {
      next = cases.__
    } else {
      console.error('no case "%s" exists', caseName)
      return
    }
    runStep(next, ctx, meta)
  },

  single(step: TypeDef<'single', 'step'>, ctx: CommonCtx, meta: Meta) {
    const single: TypeDef<*, 'cmd'> = step.data
    invariant(single.type in command, 'impossible case "%s"', single.type)
    invariant(ctx.type in command, 'impossible case "%s"', ctx.type)
    callstack.pushItem(single)
    meta.arg = command[ctx.type].stepArg(ctx.data)
    meta.ctx = command[single.type].cmd(single, ctx, meta)
    meta.stop = !command[meta.ctx.type].transition(meta.ctx)
    callstack.popItem()
  },
  multi(step, ctx, meta) {
    for (let i = 0; i < step.data.length; i++) {
      runStep(step.data[i], ctx, meta)
    }
    meta.stop = false
  },
  seq(steps: TypeDef<'seq', 'step'>, prev: CommonCtx, meta) {
    meta.ctx = prev
    for (let i = 0; i < steps.data.length; i++) {
      runStep(steps.data[i], meta.ctx, meta)
      if (meta.stop) {
        break
      }
    }
  },
}

type Command<tag> = /*:: interface */ {
  cmd(
    single: TypeDef<tag, 'cmd'>,
    ctx: CommonCtx,
    meta: Meta,
  ): TypeDef<tag, 'ctx'>,
  transition(ctx: CommonCtx): boolean,
  stepArg(data: any): any,
}

const command = ({
  emit: {
    cmd: (single, ctx, meta) =>
      Ctx.emit({
        eventName: single.data.fullName,
        payload: meta.arg,
      }),
    transition: ctx => true,
    stepArg: data => data.payload,
  },
  filter: {
    cmd(single, ctx, meta) {
      const arg = meta.arg

      let isChanged = false
      try {
        isChanged = single.data.filter(arg, ctx)
      } catch (err) {
        console.error(err)
      }
      return Ctx.filter({
        value: arg,
        isChanged,
      })
    },
    transition: ctx => Boolean(ctx.data.isChanged),
    stepArg: data => data.value,
  },
  run: {
    cmd(single, ctx, meta) {
      const arg = meta.arg

      if ('transactionContext' in single.data)
        meta.transactions.push(single.data.transactionContext(arg))
      try {
        single.data.runner(arg)
      } catch (err) {
        console.error(err)
      }
      return Ctx.run({
        args: [arg],
        parentContext: ctx,
      })
    },
    transition: ctx => false,
    stepArg: data => invariant(false, 'RunContext is not supported'),
  },
  update: {
    cmd(single, ctx, meta) {
      single.data.store.current = meta.arg
      return Ctx.update({value: meta.arg})
    },
    transition: ctx => true,
    stepArg: data => data.value,
  },
  compute: {
    cmd(single, ctx, meta) {
      const arg = meta.arg

      const newCtx = Ctx.compute({
        args: [undefined, arg, ctx],
        result: null,
        error: null,
        isError: false,
        isNone: true,
        isChanged: true,
      })
      try {
        const result = single.data.fn(undefined, arg, newCtx)
        newCtx.data.result = result
        newCtx.data.isNone = result === undefined
      } catch (err) {
        console.error(err)
        newCtx.data.isError = true
        newCtx.data.error = err
        newCtx.data.isChanged = false
      }
      return newCtx
    },
    transition: ctx => Boolean(ctx.data.isChanged),
    stepArg: data => data.result,
  },
}: {
  emit: Command<'emit'>,
  filter: Command<'filter'>,
  run: Command<'run'>,
  update: Command<'update'>,
  compute: Command<'compute'>,
})
