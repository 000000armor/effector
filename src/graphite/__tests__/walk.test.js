//@flow
//@jsx fx
//eslint-disable-next-line no-unused-vars
import fx from 'effector/stdlib/fx'
import {show} from 'effector/fixtures/showstep'
import {Ctx} from 'effector/graphite/typedef'
import {createStateRef} from 'effector/stdlib/stateref'

import {createEvent, forward} from 'effector/event'
import {walkNode} from '../walk'

function eventCtx(payload) {
  return Ctx.emit({
    __stepArg: payload,
  })
}
const Next = () => (
  <multi>
    <run
      runner={(...args) => {
        console.log(...args)
      }}
    />
  </multi>
)
describe('filter node will throw', () => {
  it('throw in a single node', () => {
    const next = Next()
    const exec = (
      <seq>
        <emit fullName="emit before filter node will throw" />
        <filter
          filter={() => {
            throw new Error('[expected error]')
          }}
        />
        <multi>
          <emit fullName="emit after filter node did throw" />
        </multi>
        {next}
      </seq>
    )
    const e2 = createEvent('e2')
    e2.graphite = {seq: exec, next}
    walkNode(e2.graphite.seq, eventCtx('e2()'))
  })
  it('throw in a multi node', () => {
    const next = Next()
    const exec = (
      <seq>
        <emit fullName="emit before filter node will throw" />
        <multi>
          <filter
            filter={() => {
              throw new Error('[expected error]')
            }}
          />
          <emit fullName="emit after filter node did throw" />
        </multi>
        {next}
      </seq>
    )
    const e3 = createEvent('e3')
    e3.graphite = {seq: exec, next}
    walkNode(e3.graphite.seq, eventCtx('e3()'))
  })
})
test('walk after multi', () => {
  const e1 = createEvent('e1')
  const next = Next()
  const exec = (
    <seq>
      <emit fullName="emit before multi" />
      <multi />
      <seq />
      <multi>
        <emit fullName="emit after multi" />
      </multi>
      {next}
    </seq>
  )
  e1.graphite = {seq: exec, next}
  console.log(exec)
  const ctx = eventCtx('e1()')
  walkNode(e1.graphite.seq, ctx)
  console.log(ctx)
})

describe('<run /> execution cases', () => {
  it('[a] tested correctly', () => {
    const fn = jest.fn()
    const trigger = createEvent('[run][a]')
    const next = Next()
    const exec = (
      <seq>
        <emit fullName="[a] tested correctly" />
        <run runner={fn} />
        {next}
      </seq>
    )
    trigger.graphite = {seq: exec, next}
    walkNode(exec, eventCtx('[run][a]()'))
    expect(fn).toHaveBeenCalledTimes(1)
    expect(fn).toBeCalledWith('[run][a]()')
  })
  it(`
| not return anything from <run/>
| run is always last step in the branch`, () => {
    const fn1 = jest.fn()
    const fn2 = jest.fn()
    const trigger = createEvent('[run][b]')
    const next = Next()
    const exec = (
      <seq>
        <emit fullName="[b] tested correctly" />
        <run runner={fn1} />
        <run runner={fn2} />
        {next}
      </seq>
    )
    trigger.graphite = {seq: exec, next}
    walkNode(exec, eventCtx('[run][b]()'))
    expect(fn1).toHaveBeenCalledTimes(1)
    expect(fn1).toBeCalledWith('[run][b]()')
    expect(fn2).not.toHaveBeenCalled()
  })
})
describe.only('<loop/> execution cases', () => {
  const VAL = 'current'
  function values(val) {
    const result = {}
    for (const key in val) {
      result[key] = val[key][VAL]
    }
    return result
  }
  function using(name, reset: any) {
    if (typeof reset === 'undefined') return {name}
    return {name, reset}
  }
  it('[a] should throw error after loop timeout', () => {
    const runner = createEvent('loop runner')
    forward({
      from: runner,
      to: {
        graphite: {
          seq: (
            <loop
              iterator={<multi />}
              branch={<run runner={() => {}} />}
              source={createStateRef(['a'])}
              until={using('until', true)}
              selector={using('select', 0)}
              item={using('item', 'initial item')}
            />
          ),
        },
      },
    })
    expect(() => runner('x')).toThrowError('infinite loop protection')
  })
  it('[b] at least not dangerous', () => {
    //   {
    //   branch: Fun,
    //   iterator: Fun,
    //   source: StateRef,
    //   until: Using,
    //   selector: Using,
    //   item: Using,
    // }
    const fnItem = jest.fn()

    const state = createStateRef(['a', 'b', 'c'])

    const iterator = (
      <seq>
        <compute fn={(arg, val) => val.counter[VAL] + 1} />
        <update val="counter" />
        <update val="select" />
        <compute fn={(arg, val) => val.counter[VAL] < val.max[VAL]} />
        <update val="until" />
      </seq>
    )
    const branch = (
      <seq>
        <compute
          fn={(arg, val) => {
            console.log('compute branch', arg, values(val))
            return val.item[VAL]
          }}
        />
        <run
          runner={e => {
            console.log('run branch', e)
            fnItem(e)
          }}
        />
      </seq>
    )
    const exec = (
      <seq>
        <compute fn={() => state[VAL].length} />
        <update val="max" />
        <compute fn={() => 0} />
        <update val="counter" />
        <loop
          iterator={iterator}
          branch={branch}
          source={state}
          until={using('until', true)}
          selector={using('select', 0)}
          item={using('item', 'initial item')}
        />
      </seq>
    )
    const runner = createEvent('loop runner')
    forward({
      from: runner,
      to: {
        graphite: {
          seq: exec,
        },
      },
    })
    runner('x')
    expect(fnItem.mock.calls).toEqual([['a'], ['b'], ['c']])
  })
})
describe('<choose /> execution cases', () => {
  it('[a] at least not dangerous', () => {
    const fnNext = jest.fn()
    const fnFoo = jest.fn()
    const fnBar = jest.fn()
    const trigger = createEvent('[choose][a]')
    const next = Next()
    const state = createStateRef('foo')
    // <single>
    //       <run runner={fnSelector} />
    //     </single>
    const selector = (
      <seq>
        <compute fn={() => 'bar'} />
        <update store={state} />
      </seq>
    )
    const caseFoo = (
      <seq>
        <compute
          fn={arg => {
            fnFoo(arg)
            return 'case foo'
          }}
        />
      </seq>
    )
    const caseBar = (
      <seq>
        <compute
          fn={arg => {
            fnBar(arg)
            return 'case bar'
          }}
        />
      </seq>
    )
    const exec = (
      <seq>
        <emit fullName="[a] at least not dangerous" />
        <choose
          state={state}
          selector={selector}
          cases={{foo: caseFoo, bar: caseBar}}
        />
        <run runner={fnNext} />
        {next}
      </seq>
    )
    trigger.graphite = {seq: exec, next}
    walkNode(exec, eventCtx('[choose][a]()'))
    expect(fnFoo).not.toHaveBeenCalled()
    expect(fnBar).toHaveBeenCalledTimes(1)
    expect(fnBar).toBeCalledWith('[choose][a]()')
    // expect(fnNext).toHaveBeenCalledTimes(1)
    // expect(fnNext).toBeCalledWith('case bar')
    expect(show(exec)).toMatchSnapshot()
    expect(exec).toMatchSnapshot()
  })
})

describe('<filter /> execution cases', () => {
  it('[a] tested correctly', () => {
    const fn = jest.fn()
    const trigger = createEvent('[filter][a]')
    const next = Next()
    const exec = (
      <seq>
        <emit fullName="[a] tested correctly" />
        <filter filter={fn} />
        {next}
      </seq>
    )
    trigger.graphite = {seq: exec, next}
    walkNode(exec, eventCtx('[filter][a]()'))
    expect(fn).toHaveBeenCalledTimes(1)
    expect(fn).toBeCalledWith('[filter][a]()')
  })
  it('[b] return changes on resolve', () => {
    const fn1 = jest.fn(() => true)
    const fn2 = jest.fn()
    const trigger = createEvent('[filter][b]')
    const next = Next()
    const exec = (
      <seq>
        <emit fullName="[b] tested correctly" />
        <filter filter={fn1} />
        <run runner={fn2} />
        {next}
      </seq>
    )
    trigger.graphite = {seq: exec, next}
    walkNode(exec, eventCtx('[filter][b]()'))
    expect(fn1).toHaveBeenCalledTimes(1)
    expect(fn1).toBeCalledWith('[filter][b]()')
    expect(fn2).toHaveBeenCalledTimes(1)
    expect(fn2).toBeCalledWith('[filter][b]()')
  })
  it('[c] not return anything on reject', () => {
    const fn1 = jest.fn(() => false)
    const fn2 = jest.fn()
    const trigger = createEvent('[filter][c]')
    const next = Next()
    const exec = (
      <seq>
        <emit fullName="[c] tested correctly" />
        <filter filter={fn1} />
        <run runner={fn2} />
        {next}
      </seq>
    )
    trigger.graphite = {seq: exec, next}
    walkNode(exec, eventCtx('[filter][c]()'))
    expect(fn1).toHaveBeenCalledTimes(1)
    expect(fn1).toBeCalledWith('[filter][c]()')
    expect(fn2).not.toHaveBeenCalled()
  })
  it('[d] not return anything on throw', () => {
    const fn1 = () => {
      throw new Error('[expected error]')
    }
    const fn2 = jest.fn()
    const trigger = createEvent('[filter][d]')
    const next = Next()
    const exec = (
      <seq>
        <emit fullName="[d] tested correctly" />
        <filter filter={fn1} />
        <run runner={fn2} />
        {next}
      </seq>
    )
    trigger.graphite = {seq: exec, next}
    walkNode(exec, eventCtx('[filter][d]()'))
    expect(fn2).not.toHaveBeenCalled()
  })
})
