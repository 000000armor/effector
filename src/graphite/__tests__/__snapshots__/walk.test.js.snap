// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<choose /> execution cases [a] at least not dangerous 1`] = `
"Seq [
  1. emit {{
     subtype: event,
     fullName: [a] at least not dangerous
  }},
  2. choose {{
     ref: {
       id: 1,
       current: bar
     },
     selector: {
       id: 2t,
       type: seq,
       group: step,
       data: [
         {
           id: 2q,
           type: single,
           group: step,
           data: {
             id: 2p,
             type: compute,
             group: cmd,
             data: {}
           }
         },
         {
           id: 2s,
           type: single,
           group: step,
           data: {
             id: 2r,
             type: update,
             group: cmd,
             data: {
               store: {
                 id: 1,
                 current: bar
               }
             }
           }
         }
       ]
     },
     cases: {
       foo: {
         id: 2w,
         type: seq,
         group: step,
         data: [
           {
             id: 2v,
             type: single,
             group: step,
             data: {
               id: 2u,
               type: compute,
               group: cmd,
               data: {}
             }
           }
         ]
       },
       bar: {
         id: 2z,
         type: seq,
         group: step,
         data: [
           {
             id: 2y,
             type: single,
             group: step,
             data: {
               id: 2x,
               type: compute,
               group: cmd,
               data: {}
             }
           }
         ]
       }
     }
  }},
  3. run {{}},
  4. [
     *  run {{}}
  ]
]"
`;

exports[`<choose /> execution cases [a] at least not dangerous 2`] = `
Object {
  "data": Array [
    Object {
      "data": Object {
        "data": Object {
          "fullName": "[a] at least not dangerous",
          "subtype": "event",
        },
        "group": "cmd",
        "id": "30",
        "type": "emit",
      },
      "group": "step",
      "id": "31",
      "type": "single",
    },
    Object {
      "data": Object {
        "cases": Object {
          "bar": Object {
            "data": Array [
              Object {
                "data": Object {
                  "data": Object {
                    "reduce": [Function],
                  },
                  "group": "cmd",
                  "id": "2x",
                  "type": "compute",
                },
                "group": "step",
                "id": "2y",
                "type": "single",
              },
            ],
            "group": "step",
            "id": "2z",
            "type": "seq",
          },
          "foo": Object {
            "data": Array [
              Object {
                "data": Object {
                  "data": Object {
                    "reduce": [Function],
                  },
                  "group": "cmd",
                  "id": "2u",
                  "type": "compute",
                },
                "group": "step",
                "id": "2v",
                "type": "single",
              },
            ],
            "group": "step",
            "id": "2w",
            "type": "seq",
          },
        },
        "ref": Object {
          "current": "bar",
          "id": "1",
        },
        "selector": Object {
          "data": Array [
            Object {
              "data": Object {
                "data": Object {
                  "reduce": [Function],
                },
                "group": "cmd",
                "id": "2p",
                "type": "compute",
              },
              "group": "step",
              "id": "2q",
              "type": "single",
            },
            Object {
              "data": Object {
                "data": Object {
                  "store": Object {
                    "current": "bar",
                    "id": "1",
                  },
                },
                "group": "cmd",
                "id": "2r",
                "type": "update",
              },
              "group": "step",
              "id": "2s",
              "type": "single",
            },
          ],
          "group": "step",
          "id": "2t",
          "type": "seq",
        },
      },
      "group": "step",
      "id": "32",
      "type": "choose",
    },
    Object {
      "data": Object {
        "data": Object {
          "runner": [MockFunction] {
            "calls": Array [
              Array [
                "case bar",
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          },
        },
        "group": "cmd",
        "id": "33",
        "type": "run",
      },
      "group": "step",
      "id": "34",
      "type": "single",
    },
    Object {
      "data": Array [
        Object {
          "data": Object {
            "data": Object {
              "runner": [Function],
            },
            "group": "cmd",
            "id": "2m",
            "type": "run",
          },
          "group": "step",
          "id": "2n",
          "type": "single",
        },
      ],
      "group": "step",
      "id": "2o",
      "type": "multi",
    },
  ],
  "group": "step",
  "id": "35",
  "type": "seq",
}
`;
